### ISSUE TITLE
Using "segmentation_models_pytorch" with Unet++

### ISSUE THREAD

**Issue by alqurri77** [isaaccorley/torchseg#4]
was trying to use segmentation_models_pytorch  with Unet++. I tried to look at the example at [HuBMAP - Pytorch smp Unet++](https://www.kaggle.com/code/nayuts/hubmap-pytorch-smp-unet). My understanding Unet++ returns 4 outputs (as part of deep supervision). But in the example, I notice it is only returning one output. I think I'm missing something here.

**Comment by isaaccorley** [comment]
Hi @alqurri77, I didn't write the initial smp implementation of UNet++ or the tutorial but I think you're right that the model _should_ be able to return the intermediate outputs for deep supervision. I'll make an issue and get a PR going for this. Stay tuned!

Thanks!

**Comment by alqurri77** [comment]
Thank you very much, Issac!

**Comment by notprime** [comment]
Hi @alqurri77 ,

I'm commenting to add infos and specify what we need to be implement.
Yes, you are right, for deep supervision we have to store the intermediate outputs of all hidden layers, both from encoder and decoder, so that we can compute the discrepancy between the decoder block's output and the corresponding down-sampled version from the encoder block. 
Now, I believe there are different papers exploring different discrepancy losses, like MSE, Cross entropy or classic segmentation losses, there is not an heuristic on which one is better.
We can simply return the necessary feature maps and leave the loss computation in the hands of the user, what do you think @isaaccorley ?


A couple of references:
- [Deeply-Supervised Nets](https://arxiv.org/pdf/1409.5185.pdf) (original paper about deep supervision)
- [Deeply supervised UNet for semantic segmentation to assist dermatopathological assessment of Basal Cell Carcinoma (BCC)](https://arxiv.org/pdf/2103.03759.pdf)

**Comment by isaaccorley** [comment]
Yes, my initial thoughts for this are to have an optional arg which tells the decoder and overall model to return the encoder and intermediate decoder outputs in the forward call.

**Comment by alqurri77** [comment]
Yes, that sounds good! @notprime @isaaccorley
